File: backend/app/models/rbac.py
"""
RBAC 权限模型

基于角色的访问控制(Role-Based Access Control),提供角色/权限/菜单/数据范围管理。
- Role: 角色表,定义系统角色(普通用户/评审员/审计员/管理员/超级管理员)
- Permission: 权限点表,定义细粒度操作权限(如 review:start)
- RolePermission: 角色-权限关联表(多对多)
- UserRole: 用户-角色关联表(多对多)
- Menu: 菜单表,自引用树形结构,关联权限点控制可见性
- DataScope: 数据范围表,定义角色可访问的数据边界
"""
from sqlalchemy import (
    JSON,
    BigInteger,
    Column,
    Index,
    Integer,
    String,
    UniqueConstraint,
)

from app.core.database import Base
from app.models.base import IdMixin, TimestampMixin


class Role(Base, IdMixin, TimestampMixin):
    """角色表

    定义系统角色,预置角色(user/reviewer/auditor/admin/super_admin)不可删除。

    Attributes:
        name: 角色名称(如"评审员")
        code: 角色编码(唯一,如 reviewer)
        description: 角色描述
        status: 状态(active/disabled)
        sort: 排序值,越小越靠前
        is_builtin: 是否预置角色(1=预置不可删,0=自定义)
    """

    __tablename__ = "role"
    __table_args__ = (
        UniqueConstraint("code", name="uk_role_code"),
        Index("ix_role_status", "status"),
    )

    name = Column(String(64), nullable=False, comment="角色名称")
    code = Column(String(64), nullable=False, comment="角色编码")
    description = Column(String(255), nullable=True, comment="角色描述")
    status = Column(
        String(16),
        nullable=False,
        default="active",
        server_default="active",
        comment="状态: active/disabled",
    )
    sort = Column(Integer, nullable=False, default=100, server_default="100", comment="排序值")
    is_builtin = Column(
        Integer,
        nullable=False,
        default=0,
        server_default="0",
        comment="是否预置角色: 1=是,0=否",
    )


class Permission(Base, IdMixin, TimestampMixin):
    """权限点表

    定义细粒度操作权限,按模块分组(如 review:start 属于 review 模块)。

    Attributes:
        code: 权限编码(唯一,如 review:start)
        name: 权限名称(如"启动审查")
        module: 所属模块(如 review)
        type: 权限类型(api/menu/button)
        description: 权限描述
    """

    __tablename__ = "permission"
    __table_args__ = (
        UniqueConstraint("code", name="uk_permission_code"),
        Index("ix_permission_module", "module"),
    )

    code = Column(String(64), nullable=False, comment="权限编码")
    name = Column(String(128), nullable=False, comment="权限名称")
    module = Column(String(32), nullable=False, comment="所属模块")
    type = Column(
        String(16),
        nullable=False,
        default="api",
        server_default="api",
        comment="权限类型: api/menu/button",
    )
    description = Column(String(255), nullable=True, comment="权限描述")


class RolePermission(Base, IdMixin, TimestampMixin):
    """角色-权限关联表

    建立角色与权限的多对多关系。

    Attributes:
        role_id: 角色ID
        permission_id: 权限ID
    """

    __tablename__ = "role_permission"
    __table_args__ = (
        UniqueConstraint("role_id", "permission_id", name="uk_role_permission"),
        Index("ix_role_permission_role", "role_id"),
        Index("ix_role_permission_perm", "permission_id"),
    )

    role_id = Column(BigInteger, nullable=False, comment="角色ID")
    permission_id = Column(BigInteger, nullable=False, comment="权限ID")


class UserRole(Base, IdMixin, TimestampMixin):
    """用户-角色关联表

    建立用户与角色的多对多关系。

    Attributes:
        user_id: 用户ID
        role_id: 角色ID
    """

    __tablename__ = "user_role"
    __table_args__ = (
        UniqueConstraint("user_id", "role_id", name="uk_user_role"),
        Index("ix_user_role_user", "user_id"),
        Index("ix_user_role_role", "role_id"),
    )

    user_id = Column(BigInteger, nullable=False, comment="用户ID")
    role_id = Column(BigInteger, nullable=False, comment="角色ID")


class Menu(Base, IdMixin, TimestampMixin):
    """菜单表

    自引用树形结构,通过 parent_id 建立父子关系,顶级菜单 parent_id 为 None。
    通过 permission_code 关联权限点控制菜单可见性。

    Attributes:
        parent_id: 父菜单ID,顶级菜单为 None
        name: 菜单名称
        path: 前端路由路径
        component: 前端组件路径
        icon: 菜单图标
        sort: 排序值
        permission_code: 关联权限编码
        visible: 是否可见(1=可见,0=隐藏)
        is_builtin: 是否预置菜单(1=预置不可删,0=自定义)
    """

    __tablename__ = "menu"
    __table_args__ = (
        Index("ix_menu_parent", "parent_id"),
    )

    parent_id = Column(BigInteger, nullable=True, comment="父菜单ID,顶级为NULL")
    name = Column(String(64), nullable=False, comment="菜单名称")
    path = Column(String(255), nullable=True, comment="前端路由路径")
    component = Column(String(255), nullable=True, comment="前端组件路径")
    icon = Column(String(64), nullable=True, comment="菜单图标")
    sort = Column(Integer, nullable=False, default=100, server_default="100", comment="排序值")
    permission_code = Column(String(64), nullable=True, comment="关联权限编码")
    visible = Column(
        Integer,
        nullable=False,
        default=1,
        server_default="1",
        comment="是否可见: 1=可见,0=隐藏",
    )
    is_builtin = Column(
        Integer,
        nullable=False,
        default=0,
        server_default="0",
        comment="是否预置菜单: 1=是,0=否",
    )


class DataScope(Base, IdMixin, TimestampMixin):
    """数据范围表

    定义角色可访问的数据边界,scope_type 控制:
    - all: 全部数据
    - project_own: 仅自己创建的项目
    - project_member: 参与的项目
    - custom: 自定义项目列表(project_ids 字段)

    Attributes:
        role_id: 角色ID
        scope_type: 范围类型(all/project_own/project_member/custom)
        project_ids: 自定义项目ID列表(custom 类型时使用)
    """

    __tablename__ = "data_scope"
    __table_args__ = (
        Index("ix_data_scope_role", "role_id"),
    )

    role_id = Column(BigInteger, nullable=False, comment="角色ID")
    scope_type = Column(
        String(32),
        nullable=False,
        comment="范围类型: all/project_own/project_member/custom",
    )
    project_ids = Column(JSON, nullable=True, comment="自定义项目ID列表(custom类型时使用)")

File: frontend/src/types/rbac.ts
/**
 * RBAC 权限模型类型定义
 * 与后端 T10 RBAC schema 对齐,涵盖角色/权限点/菜单/数据范围
 */

/**
 * 权限点定义
 */
export interface Permission {
  /** 权限点 ID */
  id: number
  /** 权限点编码,格式为 module:action,例如 project:view */
  code: string
  /** 权限点名称 */
  name: string
  /** 所属模块: project/file/review/issue/agent/report/audit/user/role/menu */
  module: string
  /** 权限点描述 */
  description?: string
}

/**
 * 角色定义
 */
export interface Role {
  /** 角色 ID */
  id: number
  /** 角色编码,例如 admin/reviewer/user */
  code: string
  /** 角色名称 */
  name: string
  /** 角色描述 */
  description?: string
  /** 是否为内置角色(内置角色不允许删除) */
  is_builtin: boolean
  /** 角色拥有的权限点列表(查询角色权限时返回) */
  permissions?: Permission[]
}

/**
 * 菜单定义(树形结构)
 */
export interface Menu {
  /** 菜单 ID */
  id: number
  /** 父菜单 ID,根菜单为 0 或 null */
  parent_id: number | null
  /** 菜单编码 */
  code: string
  /** 菜单名称 */
  name: string
  /** 菜单路由路径 */
  path: string
  /** 菜单对应组件路径 */
  component?: string
  /** 菜单图标 */
  icon?: string
  /** 排序序号 */
  sort_order: number
  /** 是否可见 */
  is_visible: boolean
  /** 子菜单列表 */
  children?: Menu[]
}

/**
 * 数据范围类型
 * - all: 全部数据
 * - project: 指定项目
 * - self: 仅本人数据
 */
export type DataScopeType = 'all' | 'project' | 'self'

/**
 * 数据范围定义
 */
export interface DataScope {
  /** 数据范围 ID */
  id: number
  /** 角色 ID */
  role_id: number
  /** 范围类型: all/project/self */
  scope_type: DataScopeType
  /** 项目 ID 列表(scope_type 为 project 时有效) */
  project_ids?: number[]
}

/**
 * 用户角色分配入参
 */
export interface UserRoleAssignIn {
  /** 用户 ID */
  user_id: number
  /** 角色 ID 列表 */
  role_ids: number[]
}

/**
 * 角色创建入参
 */
export interface RoleCreateIn {
  /** 角色编码 */
  code: string
  /** 角色名称 */
  name: string
  /** 角色描述 */
  description?: string
}

/**
 * 角色更新入参
 */
export interface RoleUpdateIn {
  /** 角色名称 */
  name?: string
  /** 角色描述 */
  description?: string
}

/**
 * 角色权限分配入参
 */
export interface RolePermissionAssignIn {
  /** 权限点 ID 列表 */
  permission_ids: number[]
}

/**
 * 数据范围更新入参
 */
export interface DataScopeUpdateIn {
  /** 范围类型: all/project/self */
  scope_type: DataScopeType
  /** 项目 ID 列表(scope_type 为 project 时必填) */
  project_ids?: number[]
}

/**
 * 用户角色项(查询用户角色返回)
 */
export interface UserRoleItem {
  /** 角色 ID */
  id: number
  /** 角色编码 */
  code: string
  /** 角色名称 */
  name: string
  /** 是否为内置角色 */
  is_builtin: boolean
}

/**
 * 按角色查询用户返回的用户项
 */
export interface RoleUserItem {
  /** 用户 ID */
  id: number
  /** 用户名 */
  username: string
  /** 昵称 */
  nickname?: string
  /** 邮箱 */
  email?: string
  /** 账号状态 */
  status: number
}
